### PR TITLE
Enable NCCL zero-copy (user buffer registration) for FSDP2

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150564

In recent versions NCCL introduced support for "user buffer registration", i.e., allowing user-owned memory (such as regular PyTorch tensors) to be "registered" (pinned, page-locked, etc.) with all the various hardware (NVLink, InfiniBand, ...) in order to support zero-copy transfers and thus accelerate communication and reduce resource footprint of NCCL's kernels (which reduces contention).

This was already exposed in PyTorch through a custom allocator provided by the NCCL process group. DDP already uses this, via a memory pool to allow caching and reusing.

FSDP2 is also particularly suited to leverage user buffer registration because the buffers it passes to NCCL are allocated by FSDP2 itself, since it anyways needs to (de)interleave the parameters to/from these private buffers.

This PR adds an extra flag to FSDP2 that tells it to use the ProcessGroup allocator for these private buffers, thus allowing it to leverage NCCL zero-copy (when supported).

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 1]
```diff
@@ -80,12 +92,13 @@
     dtype: torch.dtype,
     device: torch.device,
 ) -> tuple[torch.Tensor, torch.Tensor]:
-    all_gather_output = torch.empty(
-        (all_gather_input_numel * world_size,), dtype=dtype, device=device
-    )
-    all_gather_input = all_gather_output.narrow(
-        0, all_gather_input_numel * rank, all_gather_input_numel
-    )
+    with maybe_mem_pool():
```
Comment (prateek-typoapp) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:95]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager is used multiple times within this function. Consider reusing a single context to avoid redundant setup and teardown.
```suggestion
with maybe_mem_pool():
    all_gather_output = torch.empty(
        (all_gather_input_numel * world_size,), dtype=dtype, device=device
    )
    all_gather_input = all_gather_output.narrow(
        0, all_gather_input_numel * rank, all_gather_input_numel
    )
    foreach_copy_dsts = torch.split(all_gather_input, inp_split_sizes)
    with torch.no_grad():
        torch._foreach_copy_(foreach_copy_dsts, all_gather_inputs)
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 2]
```diff
@@ -398,9 +411,10 @@
     )
     reduce_scatter_input_numel = sum(s.numel() for s in padded_unsharded_sizes)
     reduce_scatter_output_numel = reduce_scatter_input_numel // world_size
-    reduce_scatter_input = torch.empty(
-        (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
-    )
+    with maybe_mem_pool():
```
Comment (prateek-typoapp) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:414]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager is used multiple times within this function. Consider reusing a single context to avoid redundant setup and teardown.
```suggestion
with maybe_mem_pool():
    reduce_scatter_input = torch.empty(
        (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
    )
    device_handle = _get_device_handle(device.type)
    foreach_reduce_scatter_copy_in(unsharded_grads, reduce_scatter_input, world_size)
    current_stream = device_handle.current_stream()
    reduce_scatter_stream = current_stream if reduce_scatter_stream is None else reduce_scatter_stream
    all_reduce_input = None
    all_reduce_event = None
    with device_handle.stream(reduce_scatter_stream):
        reduce_output = reduce_scatter_input.new_empty((reduce_scatter_output_numel,))
```

Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:414]: **✨ AI Suggestion:** Similar to the previous suggestion, the `maybe_mem_pool` context manager might be redundant here.  `reduce_scatter_input` appears to be used throughout the function, suggesting it's not a temporary allocation. Removing the context manager could simplify the code and potentially improve performance.
```suggestion
reduce_scatter_input = torch.empty(
    (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
)
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 3]
```diff
@@ -410,7 +424,8 @@
     all_reduce_input = None
     all_reduce_event = None
     with device_handle.stream(reduce_scatter_stream):
-        reduce_output = reduce_scatter_input.new_empty((reduce_scatter_output_numel,))
+        with maybe_mem_pool():
```
Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:427]: **✨ AI Suggestion:** The use of `maybe_mem_pool` around the allocation of `reduce_output` might be unnecessary.  If `reduce_output` is used beyond this block, allocating it directly without the context manager could be more efficient.
```suggestion
reduce_output = reduce_scatter_input.new_empty((reduce_scatter_output_numel,))
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 1 above)
Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:95]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager might not be necessary in all cases and could introduce overhead. Consider only using it when memory is a known bottleneck, or make it configurable.
```suggestion
all_gather_output = torch.empty(
    (all_gather_input_numel * world_size,), dtype=dtype, device=device
)
all_gather_input = all_gather_output.narrow(
    0, all_gather_input_numel * rank, all_gather_input_numel
)
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 2 above)
Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:414]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager might not be necessary in all cases and could introduce overhead. Consider only using it when memory is a known bottleneck, or make it configurable.
```suggestion
reduce_scatter_input = torch.empty(
    (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
)
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 3 above)
Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:427]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager might not be necessary in all cases and could introduce overhead. Consider only using it when memory is a known bottleneck, or make it configurable.
```suggestion
reduce_output = reduce_scatter_input.new_empty((reduce_scatter_output_numel,))
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 1 above)
Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:95]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager might not be necessary in this specific case and could be removed to simplify the code.  The memory pool is intended for temporary allocations, but `all_gather_output` seems to be used for the duration of the function.  Directly allocating the tensor might be more efficient.
```suggestion
all_gather_output = torch.empty(
    (all_gather_input_numel * world_size,), dtype=dtype, device=device
)
all_gather_input = all_gather_output.narrow(
    0, all_gather_input_numel * rank, all_gather_input_numel
)
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 2 above)
Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:414]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager might not be necessary and could be removed.  The memory allocation for `reduce_scatter_input` seems to be used for the duration of the function, so using the memory pool might not provide any benefit and could introduce unnecessary overhead.
```suggestion
reduce_scatter_input = torch.empty(
    (reduce_scatter_input_numel,), dtype=reduce_dtype, device=device
)
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 3 above)
Comment (Prateikx) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:427]: **✨ AI Suggestion:** The `maybe_mem_pool` context manager might be redundant here and could be removed. The `reduce_output` tensor's lifespan seems tied to the function's scope, making the memory pool potentially ineffective.
```suggestion
reduce_output = reduce_scatter_input.new_empty((reduce_scatter_output_numel,))
```

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 4]
```diff
@@ -16,6 +19,17 @@
 from ._fsdp_param import FSDPParam, ShardedState
 
 
+MEM_POOL: Optional[torch.cuda.MemPool] = None
```
[the remark this answers is not in the document]
Comment (lw) [reply torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:22]: Yes at this stage this is just an experiment to allow me to benchmark the effect of registering tensors. However, I'm happy to already receive suggestions on how to design this API!

I would also prefer to just pass the MemPool as an argument to the constructor of FSDP, but one issue with that is that, ultimately, we need to pass that MemPool to a custom operator, and I don't think MemPools can be used as arguments for ops? Unless there's some trick I'm missing (boxing/unboxing?), I believe right now we have to pass it as a global (or thread-local).

@yifuwang also used globals in a similar unlanded PR at https://github.com/pytorch/pytorch/pull/147146, although he chose to pass an allocator callable (similar to what SymmMem does), but I think a MemPool integrates better with what PyTorch already offers.

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py [hunk 5]
```diff
@@ -162,15 +165,16 @@ def foreach_all_gather(
...
-            all_gather_inputs,
-            inp_split_sizes,
-            all_gather_input_numel,
-            world_size,
-            rank,
-            dtype,
-            device,
-        )
+        with torch.cuda.use_mem_pool(mempool) if mempool else nullcontext():
+            all_gather_input, all_gather_output = torch.ops.fsdp.all_gather_copy_in(
```
Comment (lw) [inline torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py:169]: I would have preferred to only apply the context manager to the low-level `torch.empty` calls, but in this case they are happening inside a `custom_op` and thus I wrap the whole op invocation.

It is in principle possible to pass the MemPool as an argument to the custom_op. It must be converted to a C++ custom class (see https://github.com/pytorch/pytorch/pull/151100). But, even then, some issues remain, notably in terms of dependencies between PyTorch libraries:
- the MemPool is defined by the CUDA caching allocator, which lives in c10 / cuda
- the C++ custom class tooling is defined in torch / cpu (with some helpers being in ATen / cpu)

I still haven't figured out a way to make this work in all cases. I can insist if people don't like the current compromise.

File: torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py]: just took a look. seems we are proposing API like `fully_shard(mempool=)` now. what's the advantage of letting user creating the mempool outside of fsdp instead of creating mempool inside fsdp?

besides UI, seems the only thing before landing is to decide if we need to pass mempool to copy-in custom kernels?

[the remark this answers is not in the document]
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py]: these are great work!

File: torch/distributed/fsdp/_fully_shard/_fully_shard.py [hunk 6]
```diff
@@ -91,6 +91,7 @@ def fully_shard(
     mp_policy: MixedPrecisionPolicy = MixedPrecisionPolicy(),
     offload_policy: OffloadPolicy = OffloadPolicy(),
     ignored_params: Optional[set[nn.Parameter]] = None,
+    allocate_memory_from_process_group: bool = False,
```
Comment (lw) [inline torch/distributed/fsdp/_fully_shard/_fully_shard.py]: @kwen2501 For now this change is opt-in, using the extra flag specified here. In DDP, as far as I understand, this is enabled by default. Do you confirm? If so, what do we think of enabling it by default here too? LMK!

Comment (xunnanxu) [inline torch/distributed/fsdp/_fully_shard/_fully_shard.py]: So if I understand correctly this is just for collective buffers right? Does it make sense to move this call from `fully_shard` to `FSDPModule` like

```
m: FSDPModule = fully_shard(...)
m.set_allocate_memory_from_process_group_for_comm(True)
```

to keep it consistent with other (post-ctor) apis?

This is to differentiate from the regular tensor alloc during sharding in constructor.

@lw @weifengpy @kwen2501 lmk if that makes sense?

Comment (kwen2501) [reply torch/distributed/fsdp/_fully_shard/_fully_shard.py]: In DDP, it is controlled via an environment var: `DDP_DISABLE_COMM_MEM`
https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/reducer.cpp#L1201-L1202

I initially set the feature to ON, but it caused OOM in some workloads that are very close to memory limit, so I have to set the default to OFF for now. (sad face)

Comment (kwen2501) [reply torch/distributed/fsdp/_fully_shard/_fully_shard.py]: btw, if we don't want to change the API surface, env var is a good option. (And we can flip the default value internally). If we want to expose it as an API arg for now, we can also `_` it to retain flexibility for future.

**Review by kwen2501 — APPROVED** [review]
Thank you! LGTM!!

**Comment by kwen2501** [comment]
Please fill the PR description if you don't mind?

File: torch/distributed/fsdp/_fully_shard/_fully_shard.py (hunk 6 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fully_shard.py]: moving `set_allocate_memory_from_process_group_for_comm` into FSDPModule aligns with other APIs better. Later, @xunnanxu 's change is also adding a method in FSDPModule

@lw are you open to make the change?

**Review by weifengpy — CHANGES_REQUESTED** [review]

**Review by syed-ahmed — APPROVED** [review]

File: torch/distributed/fsdp/_fully_shard/_fully_shard.py [hunk 7]
```diff
@@ -523,6 +523,22 @@ def set_unshard_in_backward(self, unshard_in_backward: bool) -> None:
         if (fsdp_param_group := state._fsdp_param_group) is not None:
             fsdp_param_group.unshard_in_backward = unshard_in_backward
 
+    def set_allocate_memory_from_process_group_for_comm(self, enable: bool) -> None:
```
Comment (kwen2501) [inline torch/distributed/fsdp/_fully_shard/_fully_shard.py:527]: What if there are multiple FSDP modules? Would users need to call `.set_...` one by one?
It seems this needs to be a class method at least. 

And user would be advised to call `FSDPModule.set_...` before doing anything, just to be safe.

**Review by weifengpy — APPROVED** [review]

File: test/distributed/_composable/fsdp/test_fully_shard_comm.py [hunk 8]
```diff
@@ -1283,5 +1285,45 @@ def test_unshard_without_lazy_init(self):
...
+        os.environ["NCCL_DEBUG_FILE"] = cls.nccl_log_dir.name + "/nccl_log"
+        super()._run(*args, **kwargs)
+
+    @skip_if_lt_x_gpu(2)
+    def test_fully_shard_mempool(self):
+        torch.manual_seed(42)
+        model_args = ModelArgs()
+        model = Transformer(model_args)
+        fully_shard_fn = functools.partial(
+            fully_shard, allocate_memory_from_process_group=True
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_comm.py]: update unit test?

File: torch/distributed/fsdp/_fully_shard/_fully_shard.py (hunk 7 above)
Comment (lw) [reply torch/distributed/fsdp/_fully_shard/_fully_shard.py:527]: I very much prefer to make this a per-instance method, rather than a global classmethod, as I think there could be value in some cases to set this differently for different modules.

I can understand however the desire to make this method "recursive". I don't have the impression that other methods in this class do that, thus I don't know how exactly to achieve it, but if you have pointers please let me know!

**Comment by Skylion007** [comment]
Will this be added to FSDP1 o-o?

**Comment by weifengpy** [comment]
> Will this be added to FSDP1 o-o?

@Skylion007 no plans. it's hard to commit to maintain zero-copy in fsdp1
